Android Intent Action Crash

## Expected Behavior
The file picker should not crash on Android devices when picking a file.

## Actual Behavior
Some Android devices will crash when attempting to launch the picker with `Intent.ActionGetContent` and `Intent.ActionOpenDocument` works more consistently in our findings.

The crash is:
`java.lang.SecurityException: Permission Denial: opening provider com.google.android.apps.docs.storagebackend.StorageBackendContentProvider from ProcessRecord{c8decebd0 6909:<package_name>/u0a206} (pid=6909, uid=10206) requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs`

## Steps to Reproduce the Problem
We're unsure at the moment if this is device specific, or related to the setup of the documents browser on the Android device itself. So far we have only witnessed this crash on a Samsung S5E (SM-T720) Tablet running Android 9

  1. Create an Android app
  2. Call `Task<FileData> PickFile(string[] allowedTypes = null)` from an `async/await` routine
  3. Select a file from the picker
  4. App crashes with full stack trace below

```
Thread+RunnableImplementor.Run ()
IRunnableInvoker.n_Run (System.IntPtr jnienv, System.IntPtr native__this)
(wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.34(intptr,intptr)
java.lang.SecurityException: Permission Denial: opening provider com.google.android.apps.docs.storagebackend.StorageBackendContentProvider from ProcessRecord{c8decebd0 6909:com.tyco.illustra_tools/u0a206} (pid=6909, uid=10206) requires that you obtain access using ACTION_OPEN_DOCUMENT or related APIs
android.os.Parcel.createException Parcel.java:1966
android.os.Parcel.readException Parcel.java:1934
android.os.Parcel.readException Parcel.java:1884
android.app.IActivityManager$Stub$Proxy.getContentProvider IActivityManager.java:4039
android.app.ActivityThread.acquireProvider ActivityThread.java:6351
android.app.ContextImpl$ApplicationContentResolver.acquireUnstableProvider ContextImpl.java:2825
android.content.ContentResolver.acquireUnstableProvider ContentResolver.java:1835
android.content.ContentResolver.openTypedAssetFileDescriptor ContentResolver.java:1449
android.content.ContentResolver.openAssetFileDescriptor ContentResolver.java:1302
android.content.ContentResolver.openInputStream ContentResolver.java:1022
mono.java.lang.RunnableImplementor.n_run(Native Method)
mono.java.lang.RunnableImplementor.run(Unknown Source:0)
android.os.Handler.handleCallback Handler.java:873
android.os.Handler.dispatchMessage Handler.java:99
android.os.Looper.loop Looper.java:214
android.app.ActivityThread.main ActivityThread.java:7080
java.lang.reflect.Method.invoke(Native Method)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run RuntimeInit.java:494
com.android.internal.os.ZygoteInit.main ZygoteInit.java:965
Caused by: android.os.RemoteException: Remote stack trace:
com.android.server.am.ActivityManagerService.getContentProviderImpl ActivityManagerService.java:15638
com.android.server.am.ActivityManagerService.getContentProviderImpl ActivityManagerService.java:15547
com.android.server.am.ActivityManagerService.getContentProvider ActivityManagerService.java:16112
android.app.IActivityManager$Stub.onTransact$getContentProvider$ IActivityManager.java:11035
android.app.IActivityManager$Stub.onTransact IActivityManager.java:295
```

## Specifications

  - NuGet package version: 2.1.34
  - Platform(s): Xamarin Native (Android)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,10 @@
 # Changelog
 
+## Version 2.1.xx
+
+Android: When a `content://` URI is returned for a picked file, the plugin now
+tries to keep permissions to the picked content (#172, #158).
+
 ## Version 2.1.34
 
 (Re-)added SourceLink support

--- a/README.md
+++ b/README.md
@@ -109,6 +109,17 @@ you access it via GetStream() afterwards.
 
 ### **IMPORTANT**
 **Android:**
+The FilePath property may contain a content URI that starts with `content://`.
+The plugin tries hard to find out an actual filename, but when it can't, the
+file can only be accessed using `GetStream()` or `DataArray`. On Android
+`ContentProvider` classes are used to share data between apps. The resource
+that is accessed may not even be a file, streamed over the internet or loaded
+from a database.
+
+The plugin also tries to get a persistable content URI that can be stored for
+later access. Be prepared that this may fail, though. Content could have been
+moved to a different location or could be deleted.
+
 The `READ_EXTERNAL_STORAGE` permission is automatically added to your Android
 app project. Starting from Android 6.0 (Marshmallow) the user is asked for the
 permission when not granted yet. When the user denies the permission,

--- a/src/Plugin.FilePicker/Android/FilePickerActivity.android.cs
+++ b/src/Plugin.FilePicker/Android/FilePickerActivity.android.cs
@@ -98,7 +98,8 @@ namespace Plugin.FilePicker
         /// </summary>
         private void StartPicker()
         {
-            var intent = new Intent(Intent.ActionGetContent);
+            var intent = new Intent(
+                Build.VERSION.SdkInt < BuildVersionCodes.Kitkat ? Intent.ActionGetContent : Intent.ActionOpenDocument);
 
             intent.SetType("*/*");
 
@@ -110,7 +111,11 @@ namespace Plugin.FilePicker
                 intent.PutExtra(Intent.ExtraMimeTypes, allowedTypes);
             }
 
-            intent.AddCategory(Intent.CategoryOpenable);
+            if (Build.VERSION.SdkInt < BuildVersionCodes.Kitkat)
+            {
+                intent.AddCategory(Intent.CategoryOpenable);
+            }
+
             try
             {
                 this.StartActivityForResult(Intent.CreateChooser(intent, "Select file"), 0);

--- a/src/Plugin.FilePicker/Android/FilePickerActivity.android.cs
+++ b/src/Plugin.FilePicker/Android/FilePickerActivity.android.cs
@@ -115,6 +115,10 @@ namespace Plugin.FilePicker
             {
                 intent.AddCategory(Intent.CategoryOpenable);
             }
+            else
+            {
+                intent.AddFlags(ActivityFlags.GrantPersistableUriPermission);
+            }
 
             try
             {
@@ -152,6 +156,13 @@ namespace Plugin.FilePicker
                     System.Diagnostics.Debug.Write(data.Data);
 
                     var uri = data.Data;
+
+                    if (Build.VERSION.SdkInt >= BuildVersionCodes.Kitkat)
+                    {
+                        context.ContentResolver.TakePersistableUriPermission(
+                            uri,
+                            ActivityFlags.GrantReadUriPermission);
+                    }
 
                     var filePath = IOUtil.GetPath(this.context, uri);
 

